gems/sorbet-runtime/test/types/final.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+require_relative '../test_helper'
+
+class Opus::Types::Test::FinalTest < Critic::Unit::UnitTest
+  after do
+    T::Private::DeclState.current.reset!
+  end
+
+  it "allows declaring an instance method as final" do
+    Class.new do
+      extend T::Sig
+      sig {void}.final
+      def foo; end
+    end
+  end
+
+  it "allows declaring a class method as final" do
+    Class.new do
+      extend T::Sig
+      sig {void}.final
+      def self.foo; end
+    end
+  end
+
+  it "forbids redefining a final instance method" do
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend T::Sig
+        sig {void}.final
+        def foo; end
+        sig {void}.final
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be redefined")
+  end
+
+  it "forbids redefining a final class method" do
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend T::Sig
+        sig {void}.final
+        def self.foo; end
+        sig {void}.final
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be redefined")
+  end
+
+  it "forbids overriding a final instance method" do
+    c = Class.new do
+      extend T::Sig
+      sig {void}.final
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new(c) do
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final class method" do
+    c = Class.new do
+      extend T::Sig
+      sig {void}.final
+      def self.foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new(c) do
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+end
```

Comment:
I think there are a lot more test cases we can think of here:

- when either the first or the second method definition doesn't have a sig at all
- when the final method is defined in a module
- when two modules are included into a class, the method isn't defined in the class, but one of the modules makes the method final
- `include` and `extend` with modules 

Would you take a second pass at stress-testing the edge cases here?

Reply:
I added more tests.